src/SFML/Audio/SoundFileReaderMp3.cpp

```diff
@@ -0,0 +1,134 @@
+////////////////////////////////////////////////////////////
+//
+// SFML - Simple and Fast Multimedia Library
+// Copyright (C) 2007-2019 Laurent Gomila (laurent@sfml-dev.org)
+//
+// This software is provided 'as-is', without any express or implied warranty.
+// In no event will the authors be held liable for any damages arising from the use of this software.
+//
+// Permission is granted to anyone to use this software for any purpose,
+// including commercial applications, and to alter it and redistribute it freely,
+// subject to the following restrictions:
+//
+// 1. The origin of this software must not be misrepresented;
+//    you must not claim that you wrote the original software.
+//    If you use this software in a product, an acknowledgment
+//    in the product documentation would be appreciated but is not required.
+//
+// 2. Altered source versions must be plainly marked as such,
+//    and must not be misrepresented as being the original software.
+//
+// 3. This notice may not be removed or altered from any source distribution.
+//
+////////////////////////////////////////////////////////////
+
+////////////////////////////////////////////////////////////
+// Headers
+////////////////////////////////////////////////////////////
+#define MINIMP3_IMPLEMENTATION // Minimp3 control define, places implementation in this file.
+#include <SFML/Audio/SoundFileReaderMp3.hpp>
+#include <SFML/System/MemoryInputStream.hpp>
+#include <SFML/System/Err.hpp>
+#include <algorithm>
+
+
+namespace
+{
+std::size_t readCallback(void* ptr, std::size_t size, void* data)
+{
+    sf::InputStream* stream = static_cast<sf::InputStream*>(data);
+    return static_cast<std::size_t>(stream->read(ptr, size));
+}
+
+int seekCallback(uint64_t offset, void* data)
+{
+    sf::InputStream* stream = static_cast<sf::InputStream*>(data);
+    sf::Int64 position = stream->seek(offset);
+    return position < 0 ? -1 : 0;
+}
+}
+
+namespace sf
+{
+namespace priv
+{
+////////////////////////////////////////////////////////////
+bool SoundFileReaderMp3::check(InputStream& stream)
+{
+    Uint8 header[10];
+    if (stream.read(header, sizeof(header)) < sizeof(header))
+        return false;
+    if (!memcmp(header, "ID3", 3) && !((header[5] & 15) || (header[6] & 0x80) || (header[7] & 0x80) || (header[8] & 0x80) || (header[9] & 0x80)))
+        return true;
+    if (hdr_valid(header))
+        return true;
+    return false;
+}
+
+
+////////////////////////////////////////////////////////////
+SoundFileReaderMp3::SoundFileReaderMp3() :
+m_numSamples(0),
+m_position(0)
+{
+    m_io.read = readCallback;
+    m_io.seek = seekCallback;
```

Comment:
The rest of `m_io` remains uninitialized, since its type is a POD (plain old data) without constructor. This might lead to UB if one of the `SoundFileReaderMp3` methods is called without a prior `open()` call, where `m_io` is fully initialized.

Also, `m_decoder` is not initialized.

Reply:
Added memset initialize. Done. Updated the PR.

Reply:
When the memory is zeroed, `seek()` and other operations are safe (i.e. fail properly)?
I don't know the implementation and assumptions of the underlying library.

Reply:
It should be fine, added tests for such case https://github.com/lieff/minimp3/commit/b18d274b998cd4406070ddc1f370f53392241af0 .
But calling open() method twice will lead to mem leak. Same for oundFileReaderOgg as I can see. So calling it twice forbidden? Or may be add protection against it?